zone/bonuses.cpp

```diff
@@ -5631,3 +5641,127 @@ void Mob::NegateSpellEffectBonuses(uint16 spell_id)
+	newbon->heroic_max_end += GetHeroicAGI() * RuleR(Character, HeroicAgilityMultiplier) / 4 * 10.0f;
+	newbon->heroic_end_regen += GetHeroicAGI() * RuleR(Character, HeroicAgilityMultiplier) / 4 / 50;
+
+	if (RuleB(Character, HeroicStatsUseDataBucketsToScale)) {
+		newbon->heroic_agi_avoidance += GetHeroicAGI() * CheckHeroicBonusesDataBuckets("HAGI-Avoidance");
+		newbon->heroic_max_end += GetHeroicAGI() * CheckHeroicBonusesDataBuckets("HAGI-MaxEndurance") / 4 * 10.0f;;
+		newbon->heroic_end_regen += GetHeroicAGI() * CheckHeroicBonusesDataBuckets("HAGI-EnduranceRegen") / 4 / 50;
+	}
+}
+
+void Mob::SetHeroicStaBonuses(StatBonuses* newbon) {
+
+	newbon->heroic_max_hp += GetHeroicSTA() * RuleR(Character, HeroicStaminaMultiplier) * 10;
+	newbon->heroic_hp_regen += GetHeroicSTA() * RuleR(Character, HeroicStaminaMultiplier) / 20;
+	newbon->heroic_max_end += GetHeroicSTA() * RuleR(Character, HeroicStaminaMultiplier) / 4 * 10.0f;
+	newbon->heroic_end_regen += GetHeroicSTA() * RuleR(Character, HeroicStaminaMultiplier) / 4 / 50;
+
+	if (RuleB(Character, HeroicStatsUseDataBucketsToScale)) {
+		newbon->heroic_max_hp += GetHeroicSTA() * CheckHeroicBonusesDataBuckets("HSTA-MaxHP") * 10;
+		newbon->heroic_hp_regen += GetHeroicSTA() * CheckHeroicBonusesDataBuckets("HSTA-HPRegen") / 20;
+		newbon->heroic_max_end += GetHeroicSTA() * CheckHeroicBonusesDataBuckets("HSTR-MaxEndurance") / 4 * 10.0f;
+		newbon->heroic_end_regen += GetHeroicSTA() * CheckHeroicBonusesDataBuckets("HSTA-EnduranceRegen") / 4 / 50;
+	}
+}
+
+void Mob::SetHeroicStrBonuses(StatBonuses* newbon) {
+
+	newbon->heroic_str_shield_ac += GetHeroicSTR() * RuleR(Character, HeroicStrengthMultiplier) / 10;
+	newbon->heroic_str_melee_damage += GetHeroicSTR() * RuleR(Character, HeroicStrengthMultiplier) / 10;
+	newbon->heroic_max_end += GetHeroicSTR() * RuleR(Character, HeroicStrengthMultiplier) / 4 * 10.0f;
+	newbon->heroic_end_regen +=  GetHeroicSTR() * RuleR(Character, HeroicStrengthMultiplier) / 4 / 50;
+
+	if (RuleB(Character, HeroicStatsUseDataBucketsToScale)) {
+		newbon->heroic_str_shield_ac += GetHeroicSTR() * CheckHeroicBonusesDataBuckets("HSTR-ShieldAC") / 10;
+		newbon->heroic_str_melee_damage += GetHeroicSTR() * CheckHeroicBonusesDataBuckets("HSTR-MeleeDamage") / 10;
+		newbon->heroic_max_end += GetHeroicSTR() * CheckHeroicBonusesDataBuckets("HSTR-MaxEndurance") / 4 * 10.0f;
+		newbon->heroic_end_regen += GetHeroicSTR() * CheckHeroicBonusesDataBuckets("HSTR-EnduranceRegen") / 4 / 50;
+	}
+}
+
+float Mob::CheckHeroicBonusesDataBuckets(std::string bucket_name) {
+
+	std::string bucket_value;
+	if (!bucket_name.empty()) {
+		const auto full_name = fmt::format(
+			"{}-{}",
+			GetBucketKey(),
+			bucket_name
+		);
+		if (IsOfClientBot()) {
+			bucket_value = IsClient() ? CastToClient()->m_client_data_buckets[full_name] : CastToBot()->m_bot_data_buckets[full_name];
+		}
+		if (bucket_value.empty() || !Strings::IsNumber(bucket_value)) {
```

Comment:
Seems like we'd want to use `IsFloat` here

Reply:
Bucket can either be a float, or integer value. I don't want to pigeon hole an operator into having to use a float, especially since we just return ```return Strings::ToFloat(bucket_value);```